### ISSUE TITLE
Remove all Keybase features from Mastodon

### ISSUE THREAD

**Issue by joenepraat** [mastodon/mastodon#13670]
### Pitch and Motivation

Because Keybase is acquired by Zoom, a large privacy disrespecting US company, Mastodon's (and all privacy loving Mastodon servers) reputation is at stake.

https://blog.zoom.us/wordpress/2020/05/07/zoom-acquires-keybase-and-announces-goal-of-developing-the-most-broadly-used-enterprise-end-to-end-encryption-offering/

So this is actually not a feature request, but an anti-feature request.

**Comment by mstdn** [comment]
I also have mixed feelings about this. I have setup a poll to see what people think about this here: https://mstdn.social/@stux/104129394342656607

**Comment by Gargron** [comment]
Not often that I get people wanting me to remove something instead of adding something, it is tempting. What you should know though is that the Keybase integration is basically a public data endpoint that only returns something when a user has initiated the proof process from Keybase's side, so there is basically no tangible risk or disadvantage from having this feature. No phoning home, no exchange of data, etc. It was also built around the assumption that the API might be implemented by other identity providers as well at some point (but it hasn't happened so far).

**Comment by freemo** [comment]
I strongly support keeping keybase as a feature. As Gargron said there is absolutely no security risk, I dont see any risk to reputation, and while I am not a fan of zoom as a company even using them and contributing your public key to their system is in no way a security risk to yourself or users.

Strong vote to keep.

**Comment by shleeable** [comment]
Anybody feel like building an ActivityPub Keybase clone?
That would solve the encrypted DM problem if we did it correctly..

Imagine if the private key was kept on a FedKey instance, and the ciphertext was stored on the mastodon instance.

You'd need to hack both to decrypt ;)

**Comment by wiktor-k** [comment]
> Anybody feel like building an ActivityPub Keybase clone?

I built a keybase social proofs system that can validate ActivityPub accounts like Keybase (minus the need for instance registration ;) ): [this is how it looks like](https://metacode.biz/openpgp/key#0x6325c3370b70177138abf3086a957c9a9a9429f7). Note that this is *only* social proofs, no encryption.

As for having encrypted ActivityPub objects that's usually not how they're used nowadays (most posts are #Public anyway) so this would be a non-trivial task. Maybe it would be better to just re-use existing solutions like Matrix or XMPP that already have working E2E encryption.

**Comment by jeroenpraat** [comment]
I understand that the API is neutral and this removal request is not about the API. It's about making it Keybase impossible to communicate with this API.

It is not only because of Mastodon's non-corporate reputation, but is also creates an precedent. What's next? Signing up using Google, Twitter or Facebook? :scream: 

Anyhow, an off-switch for server admins would be enough, so servers who want to be as non-corporate as possible can stay that way.

**Comment by freemo** [comment]
I would actually love to see SSO for google, twitter, Facebook and more
importantly gitlab and github. That would be awesome.

Leave the power to opt-out in the users hands not the moderators IMO, keep
the feature.

The ability to turn it off would be ok, but personally I dont see it as
needed. The users can judge for themselves if its an appropriate service to
use or not.

**Comment by ploum** [comment]
> I understand that the API is neutral and this removal request is not about the API. It's about making it Keybase impossible to communicate with this API.

When in doubt, giving the choice to users should always be the prefered solution. Nobody is forced to use the keybase integration. In fact, I'm pretty sure that most people are not even aware.

What you want is "to make it impossible for others to do something that yourself don't want to do".

This is an extremist point of view (not anger when saying that, just pointing out that you explicitely want to force user to adopt your view without giving them the choice).   And yes, if it is useful, why not allow someone to link its Mastodon account with a Facebook and Google account? As long as no data is given and is purely optional.

**Comment by jeroenpraat** [comment]
I see a few corporate trolls in here. Yes I want to prevent people using corporate products. If you think that's extremist, so be it. I call it radical anti-capitalist. Life with it.

**Comment by FrostKnight** [comment]
I have to agree with jeroenpraat, he may not have been nice about it, but, I do think he had a point. 

Capitalism is usually bad and is accompained by greed. And how does greed make money on the web? proprietary software, etc... 

I am glad you made an option to turn it off, but I do think such a feature is really against the spirit of free software. 

Just my two cents regarding this.

**Comment by FrostKnight** [comment]
I am not sure why you thumbed me down, but this will only steer people who love free software away from you guys.
I guess mastodon isn't as free as I thought. Sad day man...

Well either way, I hope you will change your mind someday...  

But yeah, I will likely keep you guys at an arms distance, later...

:/

**Comment by freemo** [comment]
@FrostKnight  I am the one who thumbed you down, I am not a mastodon dev (though I do maintain a branch). I thumbed you down because I disagree with your opinion, I am sorry if this made you feel bad but it probably isnt a good reason for you to be upset with the developers or the project, as I am not affiliated with either, just a user.

I will add though, I "love free software", I maintain countless open-source projects (some quite popular), and have have been a huge part of the open source community for many decades. All I can say is as a lover of open-source I believe in free as in freedom, freedom means choices, freedom means giving you the option to use a feature or not, and not trying to enforce that on others for ideological reasons.

As a lover of open source I can say the fact that they keep it in as steered me towards the project, not away, and removing it for purely ideological reasons, rather than giving me, the user, the **freedom** to decide would steer me away from the project.

"Free as in freedom"

**Comment by FrostKnight** [comment]
Well, given that zoom acquired keybase, you should know that if Zoom decides to throw weight behind their copyright to cause trouble, what then?  That being said, mastodon is open source, I agree with you there. I just don't think that keybase belongs in this repo for that reason and the others I have mentioned.  Compared to the guy above who seems even more anti capitalist then me, you may wish to know I am slightly more moderate then him in my views.  Free as in freedom usually in my experience involves removing stuff like keybase as far as people like Stallman are concerned.  I don't agree with Stallman on everything, for example, even if something is "libre" like systemd, dbus, pulseaudio, networkmanager, udev, logind, or redhat made, where it breaks existing backwards compatibility forcing adoption, then yes, it still should be removed even if it is libre.  So you might say, I have two thought processes, i dislike proprietary, I dislike sudo-libre as in like software that breaks freedom even if it has a free license.  My point being, BSD and GNU/Linux-Libre both have strengths and weaknesses. Someone really should merge the best of both to create a good operating system.   BSD's advantage is no fake freedom like redhat's systemd or dbus like software and its built for massive security, GNU/Linux-Libre focuses on more freedom friendly licenses and not requring permissive licenses to add it into their repos.
freedom requires good privacy.  IF you don't have good security, privacy is impossible. 

This was a bit of information that seemed like a rant, but yeah, if you think I am bad in this way, I have been to some forums where its unreal how much ideology they have. While simultaneously denying other problems.  Like that good freedom requires good privacy which requires good security.     Anywho these are my thoughts.  Take it or leave it I guess,  but tell me your thoughts on this if you wish.

**Comment by freemo** [comment]
> Well, given that zoom acquired keybase, you should know that if Zoom decides to throw weight behind their copyright to cause trouble, what then?

Well all open-source software is copyright of their respective owners and keybase was copyright before zoom, so im not sure if you think the copyright itself is the issue or the ownership by zoom. I presume its the ownership by zoom your concerned with. that said, depends on what "trouble" your talking about. I cant envision any way in which zoom can cause trouble that is in anyway damaging to mastodon. I can only speculate the scenarios you are warning against because you were explicit, but the few I can think of.

1) Keybase uses their copyright to deny its use in mastodon... ok, we cant use it anymore, take it out, problem solved no harm done.

2) Zoom attempts to harvest data about users.. ok, well anyone who uses keybase explicitly wants their public key to be publicly discoverable and have a clear path linking it to their mastodon account, thats the whole point of using keybase. So zoom cant harvest any information they couldnt have harvested if they werent involved anyway

3) They might introduce a backdoor into the system... Well they explicitly discourage people uploading their private keys, if you do so its at your own risk knowing full well its insecure. So unless you gave them your private key there is nothing they could possibly use a backdoor to do, they dont have and cant obtain your private key.

So yea.. as far as "what then".. i really cant think of any scenario this is a problem and the absolute worse outcome for anyone is that it is becomes a service that no longer serves us at which point it gets removed from the source, again, I see no harm  in this, we loose nothing.

> Compared to the guy above who seems even more anti capitalist then me, you may wish to know I am slightly more moderate then him in my views. 

Thank you for sharing that, but honestly I have no issue with you or the other guy nor have any desire to get into a capitalism vs socialism vs communism debate. I think you should be free to use the software to serve wherever you are on that spectrum and I should be free to do the same wherever I am.

> Free as in freedom usually in my experience involves removing stuff like keybase as far as people like Stallman are concerned.

I really dont care what Stallman's opinions of or application of that phrase might be to be honest, even if he is one of the founders of the movement. I care about the and have no wish or desire to emulate anyone or look to anyone for inspiration. I understand as the source of the phrase why you would invoke him, but his opinons or what he would do in no way should serve as a model in my opinion. I'd rather discuss the ideology as its own idea and not invoke others as the arbiter of a true scotsman (refernce to the No true scotsman fallacy)

> So you might say, I have two thought processes, i dislike proprietary...

This is obviously where the meat of your argument seems to rest, you dislike proprietary and you see it as anti-open source to in any way use proprietary tech. to some extent I agree, but like all ideologies the moment you take this to a purist level, as seems to be the case here, it ceases to be a useful ideology.

Particularly the use of keybase is a good example here. You invoke copyright concerns, but this seems ignorant of the fact that mastodon never touches or uses anything copyrighted in flow to support keybase. the whole process is quite literally done on the keybase website and has very little to do with mastodon up until the end, where it does little more than pass a link. Literally the whole interaction with a mastodon server is that keybase redirects a user to mastodon, the user logs in or clicks authorize to show that it was intentional, and mastodon just pings a keybase webhook saying the user authorized, using open http standards to secure that transaction, nothing proprietary from keybase of anysort. At this point all mastodon has is a random token identifier (again not copyright it is just a single id) and nothing more. It can use this token and call a HTTP endpoint and ask if the user is validated whenever it wants.

In other words absolutely nothing about the communication between keybase and mastodon is proprietary, nor is there any reason someone couldnt clone keybase and because the process is little more than a HTTP endpoint that takes an ID literally any implementation could, if they wish, replicate this without fear of copyright concerns.

So no we arent really talking about proprietary concerns, copyright concerns, or even a protocol that isnt libre. It is in no way comparable to removing pulseaudio or anything else out of protocol concerns, nor is it similar to not accepting proprietary software.. nothing proprietary has been added to mastodon.

To me your stance would be akin to an open source web browser blocking all commercial sites and only allowing access to open source websites out of some extreme ideological purism about everything proprietary being bad in any form and must be excluded from interaction. I'd be fine with a web browser not supporting HTTP v15.7 (fictional) which happens to be copyright and not public domain and thus excluding any websites which only talk in such a protocol to contrast it. 

As with all ideologies one must look at the nuance and understand the difference. When you make it about "all proprietary bad" then you are no longer writing free software, you are selling a personal agenda. Its no longer about the community and their freedom to choose but about your fight against everything proprietary and your insistence that anyone who uses this software must be part of that same fight with you, forced at a software level. That's not free software, that is not freedom.

> freedom requires good privacy. IF you don't have good security, privacy is impossible.

Yes I agree here. As pointed out earlier there is no compromise of security here, nor is it even possible for Keybase or zoom to violate your privacy if they wanted to. The security of keybase is entierly dependent on the security of your keys as well. In short no matter how much you may mistrust zoom, there is absolutely no way your privacy or security is in jeopardy here, particularly if you dont use the feature, but even if you do. Keybase doesnt get anything that isnt already public.

 > This was a bit of information that seemed like a rant, but yeah, if you think I am bad in this way...

I dont think your bad in this way or any other, I dont know you, and I try hard not to make assumptions about the character of people I dont know. I am only evaluating your stance on the issue. As a stance I do find it ideologically pure, and feels distorted by more of an anti-proprietary stance than from a pro-freedom stance. I can not and would not speak of your or your ideals in general outside of this one application and one opinion on keybase, or assume your opinion is a pattern in general. I dont know you enough to claim such a thing, nor would I want to make such a broad claim about someone anyway.

> Take it or leave it I guess, but tell me your thoughts on this if you wish.

Well you asked, so there you go, hope it helps explain my stance, I think I understand yours, even if i dont agree with it. While I strongly disagree with your opinion I have no issue with you, and didnt thumb down your post as an attack. We disagree, we both love libre but have different perspectives, it happens, no biggie, I did not mean nor intended anything personal. But after hearing your opinion my own has not changed.

**Comment by FrostKnight** [comment]
Well, good to know, but yeah, I think you understand this, but for those who don't, Stallman is not perfect and gets somethings wrong.  As I did say, freedom can even be broken by freely licensed software such as dbus, systemd, networkmanager, pulseaudio, etc...  But on topic of the copyright, If you think copyright issues won't happen because of this, or security issues, then I guess it doesn't matter. I tend to doubt it with proprietary software, but w/e, it doesn't really matter.  Heh...    
I am not sure you love libre software as much as I do, but I can tell you love open source software, which is definitely a giant step in the right direction. As is, your love of privacy and security.   Kind of surprised you aren't a dev here, I thought you were a contributor, well either way...  if you want to ever chat, ask me and I will email you.   That being said though,  I myself believe in a form of socialism with heavy restraints on captialism.  Which means, everyone is as close to being treated the same under the law regardless of how much money they have.  This of course is difficult to implement, maybe near impossible, but shoot for the moon always, or you will have way too many regrets.  :) 
peace...

**Comment by freemo** [comment]
> Well, good to know, but yeah, I think you understand this, but for those who don't, Stallman is not perfect and gets somethings wrong. As I did say, freedom can even be broken by freely licensed software such as dbus, systemd, networkmanager, pulseaudio, etc...

Yes agreed

> But on topic of the copyright, If you think copyright issues won't happen because of this, or security issues, then I guess it doesn't matter. I tend to doubt it with proprietary software, but w/e, it doesn't really matter. Heh...

Oh dont get me wrong I'm sure if zoom could use copyright to their advantage, even if it screws over people, they would. Its all about if they have something to gain. Its not that I dont think proprietary software can or would cause such issues. It is just that the nature of it in this case is such that even if they wanted or tried to hurt the mastodon community in some way, they cant, because we arent using their copyrighted or proprietary material on our servers in any way, we are simply pinging an end point they provide and displaying its response. the worse they can do is forbid us from using their endpoint on mastodon anymore and the consequence of that is the feature just gets removed, no harm done.

> I am not sure you love libre software as much as I do,

I do love it, its not that i love it less, I just love it for different reasons (not out of a hate for prroprietary software but out of a love for libre in its own right and what it provides, I simply see libre as superior, not proprietary as evil.

> but I can tell you love open source software, which is definitely a giant step in the right direction. As is, your love of privacy and security. Kind of surprised you aren't a dev here, I thought you were a contributor

Thank you, I appreciate that. I am not a contributor as stated and would not want to confuse people into thinking i can or do represent the mastodon team in anyway. I have my opinions, and it doesnt reflect theirs.

With that said I do maintain a fork of mastodon, of which mastodon is welcome to include my feature enhancements. But generally I implement the features that Eugene opposes.. Things like markdown support in posts which he has been adamatly against. So generally my enhancements do not make it back into mastodon. I am the creator and mainter of the QOTO fork of mastodon, our own flavor of mastodon as well as an instance on the network. You may have heard of us we are one of the larger instances. 

> well either way... if you want to ever chat, ask me and I will email you. 

I am satisfied with our resolution here, so i dont have any particular need to chat. But you are always most welcome to email me if you wish. You can also reach me on the mastodon network at @freemo@qoto.org if you wish to talk there. Chatting or just generally debating any topic that interests you is always welcome when polite and civil, so my door is open in that regard.

> That being said though, I myself believe in a form of socialism with heavy restraints on captialism. Which means, everyone is as close to being treated the same under the law regardless of how much money they have. This of course is difficult to implement, maybe near impossible, but shoot for the moon always, or you will have way too many regrets. :)
> peace...

It is hard to say if i agree here or not, it would also be off topic for an issue comment thread so I wont. But depends on the nuance. If you want to chat politics you know where to reach me. Chances are we will agree on a lot, and disagree on a lot too.

See ya, glad we resolved it, nice talking, reach out to me anytime.

**Comment by wiktor-k** [comment]
> It was also built around the assumption that the API might be implemented by other identity providers as well at some point (but it hasn't happened so far).

@Gargron, is there a documentation or a class where one would start implementing a different identity provider?

I'd like to take a look at how complex would it be to add https://keyoxide.org/ as a second option.

**Comment by nightpool** [comment]
@wiktor-k https://github.com/tootsuite/mastodon/blob/main/app/lib/proof_provider/keybase.rb contains the code to implement in Mastodon's codebase. You can use the existing Keybase API as reference (https://book.keybase.io/guides/proof-integration-guide) but as long as you provide all of the functions in that class, you should be good to go no matter what your API looks like.

**Comment by wiktor-k** [comment]
Great, thanks for the reference @nightpool!

**Comment by Saiv46** [comment]
Can I suggest https://keyoxide.org/ as a Keybase alternative?

**Comment by spitfire** [comment]
I noticed my verification in Keybase for my mastodon.social account has dropped, and found this.
What have we gained by removing this feature @Gargron? Do we have a drop in replacement with feature parity that is relatively easy to use?
Why did few (not even sure if that many) extremists get to decide? This project is now hosted on Microsoft-owned platform now (and it didn't explode yet) @joenepraat so the joke is on you.
What I see is an arbitrary decision (which is definitely not in open-source spirit) based on someone's political beliefs.
